### ISSUE TITLE
Mesa broke rendering in Chromium/Chrome/Electron

### ISSUE THREAD

**Issue by RaitoBezarius** [NixOS/nixpkgs#244742]
Heads-up, Nixpkgs unstable got itself into a https://bugzilla.redhat.com/show_bug.cgi?id=2193335 situation.

Therefore, you may experience garbled fonts in your program, to solve this, you can follow the instructions in https://discussion.fedoraproject.org/t/chromium-based-browsers-display-garbled-web-pages-after-mesa-is-updated/83438 option A) or option B).

Option C) is also possible, though, I do not advise it as Flatpak bypasses classical Nix semantics.

**Comment by lorenz** [comment]
Yesterday I also observed this issue on 23.05 stable, so it does not seem to be exclusive to unstable.

**Comment by RaitoBezarius** [comment]
> Yesterday I also observed this issue on 23.05 stable, so it does not seem to be exclusive to unstable.

Thanks, fixed it.

**Comment by anoadragon453** [comment]
> Therefore, you may experience garbled fonts in your program, to solve this, you can follow the instructions in [discussion.fedoraproject.org/t/chromium-based-browsers-display-garbled-web-pages-after-mesa-is-updated/83438](https://discussion.fedoraproject.org/t/chromium-based-browsers-display-garbled-web-pages-after-mesa-is-updated/83438) option A) or option B).

Option B works after restarting the program. But after a reboot I find the problem resurfaces.

**Comment by SimonBrandner** [comment]
Yeah... https://github.com/NixOS/nixpkgs/issues/242113 seems to be a duplicate of this

(also, hey @anoadragon453 :) )

**Comment by cleeyv** [comment]
In the past week there was a mesa issue opened for a similar bug: https://gitlab.freedesktop.org/mesa/mesa/-/issues/9761 I would have expected something was reported there earlier but I couldn't find anything. 

For a workaround, I have been using the command line flag `--disable-gpu-shader-disk-cache` which for both electron-based applications and chromium-based browsers has resolved the problem without completely disabling all GPU hardware acceleration as recommended elsewhere.

For element-desktop:
```
nixpkgs.config.packageOverrides = pkgs: {
  element-desktop = pkgs.element-desktop.overrideAttrs (old: {
    desktopItem = old.desktopItem.override (old: {
      exec = "element-desktop --disable-gpu-shader-disk-cache %u";
    });
  });
};
```

For brave browser:
```
nixpkgs.overlays = [
  (self: super: {
    brave = super.brave.override {
      commandLineArgs = "--disable-gpu-shader-disk-cache";
    };
  })
];
```

**Comment by SuperSandro2000** [comment]
> option B).

boils down to running ``rm -rf ~/.config/*/{Default/GPUCache,GrShaderCache,ShaderCache}`` and for simplicity just delete it for everything chromium based.

**Comment by SuperSandro2000** [comment]
I think Ubuntu worked around this by including the build version in some internal mesa version string. see https://bugs.launchpad.net/ubuntu/+source/mesa/+bug/2020604

I don't think that would work for us and I am mostly guessing from the thread because I couldn't find the actual code change in ***** launchpad.

**Comment by SuperSandro2000** [comment]
Why was this issue unpinned? This continues to be an issue for many people (encountered it twice in the last hour on different peoples laptops) especially when staging-next gets merged.

**Comment by RaitoBezarius** [comment]
The release cycle will soon start and @figsoda issue will have priority on that.

In the meantime, the 2 security vulns ongoing are much worse than broken rendering and we have plenty of folks who can recognize this issue by now.

The limit of pins being 3, here we are.

**Comment by SuperSandro2000** [comment]
> In the meantime, the 2 security vulns ongoing are much worse than broken rendering and we have plenty of folks who can recognize this issue by now.

This isn't matching my experience. Very few people around me talked about these CVEs compared to having broken chromium based programs.

> Therefore, you may experience garbled fonts in your program

I also just noticed that this isn't true for all cases. Every case I experienced myself or saw on other peoples devices, the application was completely blank and most UI elements inside the window where not rendered.

**Comment by RaitoBezarius** [comment]
> > In the meantime, the 2 security vulns ongoing are much worse than broken rendering and we have plenty of folks who can recognize this issue by now.
> 
> This isn't matching my experience. Very few people around me talked about these CVEs compared to having broken chromium based programs.

Right, but, this is not how we do our prioritization, if you feel like, this is wrong, please bring evidence in the appropriate venues to discuss how we can generalize and ground our decisions based on your experience.

> > Therefore, you may experience garbled fonts in your program
> 
> I also just noticed that this isn't true for all cases. Every case I experienced myself or saw on other peoples devices, the application was completely blank and most UI elements inside the window where not rendered. So the applications where not usable at all which is a big blocker for the next stable release.

Are you certain this is the same bug or not another issue with Electron ongoing? I also noticed this, but this is different IME.

**Comment by SuperSandro2000** [comment]
> Are you certain this is the same bug or not another issue with Electron ongoing? I also noticed this, but this is different IME.

I think this is the same issue. The fedora thread links to [this thread](https://discussion.fedoraproject.org/t/latest-fedora-update-broke-google-chrome-and-some-extensions-ui/83295/3) and [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1442633) which is very similar that it feels like the same issues I have seen. Also the mentioned workarounds like deleting the Cache directories fixed the issue everywhere.

**Comment by Xyz00777** [comment]
i have also a problem with my system, propably connected to these, while i used 23.11 pre build channel all applications worked. after switching to 24.05 pre build channel multiple applications are not rendering normal anymore. As example Obsidian, but it didnt worked with `obsidian --disable-gpu` instead it worked on my system with `rm -rf ~/.config/obsidian/GPUCache/` and than starting it, fixed my problam with obsidian but not with example bitwarden, because it dont have the simmilar folder like obsidian, is it connected?

**Comment by AmeerTaweel** [comment]
> Boils down to running `rm -rf ~/.config/*/{Default/GPUCache,GrShaderCache,ShaderCache}` and for simplicity just delete it for everything chromium based.

This worked for me on Brave Browser. However it deleted my settings :smiling_face_with_tear:. Make sure to backup/sync your settings/open tabs.

**Comment by SuperSandro2000** [comment]
I am not sure how that could happen.

**Comment by PetarKirov** [comment]
For those interested in the technical details, I really liked the summary made by Tom in the redhat bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=2193335#c41

**Comment by PetarKirov** [comment]
Yeah, I think 121.0.6100.0 is the first version that includes the fix:
https://chromium.googlesource.com/chromium/src/+/refs/tags/121.0.6100.0/gpu/command_buffer/service/shared_context_state.cc#106

According to https://chromiumdash.appspot.com/releases?platform=Linux (not sure if this a canonical or at least reliable source) the dev channel already includes the fix as of Nov 6 2023 (and there are 3 new releases since then).

**Comment by raspher** [comment]
Similar problem reappeared with brave. I've ran `brave --disable-gpu` and it launched, then in brave settings disabled hardware acceleration. After that brave is launching normally (but without acceleration).

**Comment by emilylange** [comment]
Closing, since this has been fixed ages ago.
